File: Part 1 - Data Preprocessing/Section 2 -------------------- Part 1 - Data Preprocessing --------------------/data_preprocessing_template.R
# Importing the dataset
dataset = read.csv("Data.csv")

# Taking care of missing data
# mean é uma função do R onde na.rm é um parâmetro que remove todos os NA (NaN) antes de calcular a média
# a funcao ifelse recebe condição, return se true, return se false
# se a coluna tiver um NA vai dar true e vai retornar a média dessa coluna ignorando os NA
dataset$Age = ifelse(is.na(dataset$Age),
                     ave(dataset$Age, FUN = function(x) mean(x, na.rm = TRUE)),
                     dataset$Age)
dataset$Salary = ifelse(is.na(dataset$Salary),
                     ave(dataset$Salary, FUN = function(x) mean(x, na.rm = TRUE)),
                     dataset$Salary)

# Encoding categorical data
# c é um vetor no R
# com esse processo a coluna assume valores do tipo factor, e não númericos, apesar de serem números
dataset$Country = factor(dataset$Country,
                        levels = c("France","Spain","Germany"),
                        labels = c(1,2,3))
dataset$Purchased = factor(dataset$Purchased,
                         levels = c("No","Yes"),
                         labels = c(1,2))

# Splitting the dataset in train set and test set
# SplitRatio se refere ao train
# library é um modo de importar um package via código, não necessitando clicar nele no rstudio
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = 0.8)
train_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature scaling
# ao contrário do vídeo eu tranformei em numeric Country fiz o scale para ficar igual o que foi feito em python
train_set$Country = as.numeric(train_set$Country)
test_set$Country = as.numeric(test_set$Country)
train_set[,1:3] = scale(train_set[,1:3])
test_set[,1:3]  = scale(test_set[,1:3])

File: Part 1 - Data Preprocessing/Section 2 -------------------- Part 1 - Data Preprocessing --------------------/data_preprocessing_template.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
# [linhas, colunas]
# : = todos
# :-1 = todos até ultima coluna (excluindo ela)
# -1 = somente a ultima, nessa tabela -1 seria equivalente à 3
# 1:3 = index 1 e 2
# intervalores assim incluem o primeiro limite e excluem o segundo
dataset = pd.read_csv("Data.csv")
X = dataset.iloc[:,:-1].values
Y = dataset.iloc[:,-1].values

# Taking care of missing data
# axis = 0 usa a strategy com as colunas e axis = 1 com as linhas
from sklearn.preprocessing import Imputer
imputer = Imputer(missing_values="NaN", strategy="mean", axis=0)
X[:,1:3] = imputer.fit_transform(X[:,1:3])

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:,0] = labelencoder_X.fit_transform(X[:,0])
# categorical_features já faz referencia que usaremos a coluna de index 0
# esse processo transforma a variável em dummy variables, uma variavel representada por várias colunas, onde cada coluna em 1 representa que esse estado é o ativo
# esse preço deve ser aplicado em variaveis que um estado não assume um valor maior que o outro
# aplicavel á França, Brasil, onde um estado não vale mais que o outro
# não aplicavel á L > M > S, onde um estado vale mais que o outro
onehotencoder = OneHotEncoder(categorical_features=[0])
X = onehotencoder.fit_transform(X).toarray()
# labelencoder_X não pode ser reusado pois já deu fit em outro parâmetro
labelencoder_Y = LabelEncoder()
Y = labelencoder_Y.fit_transform(Y)

# Avoiding the Dummy Variable Trap
# remove uma das colunas das dummy variables
# a bibliotega de regressão linear do python já faz isso, mas algumas bibliotecas não fazem, então é sempre bom fazer
X = X[:,1:]

# Splitting the dataset in train set and test set
# poderia definir o train_size. Os dois juntos é redundante
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

# Feature scaling
# já que os dados não precisaram mais serem vizualizados é o momento de fazer o feature scaling e "estragar" os dados
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
# IMPORTANTE: aqui faz só o transform, pois o fit já foi feito no train, isso faz com que ambos esteja escaladas na mesma base
X_test = sc_X.transform(X_test)
# em classificações o Y não precisa ser escalado. Em regreção sim
